drivers/i2c/i2c_renesas_rx_riic.c

```diff
@@ -0,0 +1,393 @@
+
+	/* Setup pin control */
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret) {
+		LOG_ERR("Pin control config failed.");
+		return ret;
+	}
+
+	/* Init kernal object */
+	k_mutex_init(&data->bus_lock);
+	k_sem_init(&data->bus_sync, 0, 1);
+
+	/* Open Device */
+	ret = R_RIIC_Open(&data->rdp_info);
+
+	if (ret) {
+		LOG_ERR("Open i2c master failed.");
+		return -EIO;
+	}
+
+	/* Connect and enable Interrupts in zephyr */
+	config->irq_config_func(dev);
+	return 0;
+}
+
+static int i2c_rx_configure(const struct device *dev, uint32_t dev_config)
+{
+	/* Setup bitrate */
+	struct i2c_rx_data *data = (struct i2c_rx_data *const)dev->data;
+	uint16_t speed; /* bitrate in kbps */
+	int ret;
+
+	/* Validate input */
+	if (!(dev_config & I2C_MODE_CONTROLLER)) {
+		LOG_ERR("Only I2C Master mode supported.");
+		return -ENOTSUP;
+	}
+	if (dev_config & I2C_ADDR_10_BITS) {
+		LOG_ERR("Only address 7 bits supported.");
+		return -ENOTSUP;
+	}
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		speed = 100;
+		break;
+	case I2C_SPEED_FAST:
+		speed = 400;
+		break;
+	default:
+		LOG_ERR("Unsupported speed: %d", I2C_SPEED_GET(dev_config));
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->bus_lock, K_FOREVER);
+
+	ret = riic_bps_calc(&data->rdp_info, speed);
+
+	k_mutex_unlock(&data->bus_lock);
+
+	/* Validate result */
+	if (ret) {
+		return -EINVAL;
+	}
+
+	data->dev_config = dev_config;
+	return 0;
+}
+
+static int i2c_rx_get_config(const struct device *dev, uint32_t *dev_config)
+{
+	struct i2c_rx_data *data = (struct i2c_rx_data *const)dev->data;
+
+	*dev_config = data->dev_config;
+	return 0;
+}
+
+static int i2c_rx_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			   uint16_t addr)
+{
+	struct i2c_rx_data *data = (struct i2c_rx_data *const)dev->data;
+
+	data->user_callback = NULL;
+	data->user_data = NULL;
+	data->skip_callback = 1;
+
+	/* Transfer */
+	return run_rx_transfer(dev, msgs, num_msgs, addr, false);
+}
+
+#ifdef CONFIG_I2C_RENESAS_RX_CALLBACK
+static int i2c_rx_transfer_cb(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			      uint16_t addr, i2c_callback_t cb, void *userdata)
+{
+	struct i2c_rx_data *data = (struct i2c_rx_data *const)dev->data;
+
+	data->user_callback = cb;
+	data->user_data = userdata;
+	data->skip_callback = 1;
+
+	/* Transfer */
+	return run_rx_transfer(dev, msgs, num_msgs, addr, true);
+}
+#endif
+
+static int i2c_rx_recover_bus(const struct device *dev)
+{
+	LOG_DBG("Recover I2C Bus\n");
+	struct i2c_rx_data *data = (struct i2c_rx_data *const)dev->data;
+	riic_return_t rdp_ret = RIIC_SUCCESS;
+
+	k_mutex_lock(&data->bus_lock, K_FOREVER);
+	rdp_ret |= R_RIIC_Control(&data->rdp_info, RIIC_GEN_START_CON);
+	rdp_ret |= R_RIIC_Control(&data->rdp_info, RIIC_GEN_STOP_CON);
+	k_mutex_unlock(&data->bus_lock);
+
+	if (rdp_ret != 0) {
+		return -EIO;
+	}
+	return 0;
+}
+
+static const struct i2c_driver_api i2c_rx_driver_api = {
```

Comment:
Please either natively implement .submit or assign it to the fallback as other drivers have done, sensors are moving to the asynchronous design that will take advantage of this. Native implementation of submit will save firmware one or more thread stacks.